shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -2594,169 +2611,211 @@ impl KafkaSinkCluster {
-                                .unwrap_or(true)
-                        })
-                        .collect();
-                    let hash = hash_partition(topic.topic_id, partition.partition_index);
-                    let shotover_node =
-                        &shotover_nodes_in_rack[hash % shotover_nodes_in_rack.len()];
-                    partition.leader_id = shotover_node.broker_id;
-                } else {
-                    // If the cluster detects the broker is down it will not be included in the metadata list of brokers.
-                    // In that case we wont find a broker and we should just mark the leader as not existing by setting it to -1
-                    // The brokers also set the leader to -1 if they detect it is down, so this matches their behaviour.
-                    partition.leader_id = BrokerId(-1);
-                }
-
-                // Every replica node has its entire corresponding shotover rack included.
-                // Since we can set as many replica nodes as we like, we take this all out approach.
-                // This ensures that:
-                // * metadata is deterministic and therefore the same on all shotover nodes
-                // * clients evenly distribute their queries across shotover nodes
-                let mut shotover_replica_nodes = vec![];
-                for replica_node in &partition.replica_nodes {
-                    if let Some(rack) = self
+        if !up_shotover_nodes.is_empty() {
+            // Overwrite list of brokers with the list of UP shotover nodes
+            metadata.brokers = up_shotover_nodes
+                .iter()
+                .map(|shotover_node| {
+                    (
+                        shotover_node.broker_id,
+                        MetadataResponseBroker::default()
+                            .with_host(shotover_node.address.host.clone())
+                            .with_port(shotover_node.address.port)
+                            .with_rack(Some(shotover_node.rack.clone())),
+                    )
+                })
+                .collect();
+
+            // Overwrite the list of partitions to point at UP shotover nodes within the same rack if any
+            // If there is no UP shotover node within the same rack, fall back to UP shotover nodes out of the rack
+            for (_, topic) in &mut metadata.topics {
+                for partition in &mut topic.partitions {
+                    // Try to deterministically choose a single UP shotover node in the rack as leader based on topic + partition id
+                    if let Some(leader_rack) = self
                         .nodes
                         .iter()
-                        .find(|x| x.broker_id == *replica_node)
+                        .find(|x| x.broker_id == *partition.leader_id)
                         .map(|x| x.rack.clone())
                     {
-                        for shotover_node in &self.shotover_nodes {
```

Comment:
Looking closer at the original implementation, cant we just keep this and replace self.shotover_nodes with up_shotover_nodes?

It avoids the allocations from partition_shotover_nodes_by_rack and avoids an intermediate HashSet.

Reply:
But in this case, there might be an edge case where `shotover_replica_nodes` is empty because it doesn't fall back to out of rack shotover nodes. Is it possible that brokers all have rack and there's no UP shotover node in rack of each broker? If that's impossible, then we can use this approach and `shotover_replica_nodes` will never be empty

Reply:
ah I see what you mean! fair enough.